### DISCUSSION TITLE
Strange behavior with cookies in route handler

### DISCUSSION THREAD

**Discussion by benetche** [vercel/next.js#70790] [Help] [answered]
### Summary

Hello! I'm creating a NextJS project and I'm currently being smashed by cookies. 
I have this page: 
```
// src/app/dashboard/page.tsx
import { cookies } from "next/headers";
import Navbar from "../components/header";
import axios from "axios";

export default async function Dashboard() {
	await axios.get("http://localhost:3000/api/dashboard");
	console.log(cookies().getAll());
	return (
		<main className="flex flex-col items-center justify-center min-h-screen bg-black-100">
			{"... some code"}
		</main>
	);
}
```
I also have this route handler, where I'm trying to set cookies:
```
// src/app/api/dashboard/route.ts
import axios from "axios";
import { cookies } from "next/headers";
import { NextResponse } from "next/server";

interface TokenData {
	access_token: string;
	token_type: string;
	expires_in: number;
}

export async function GET(req: Request) {
	const clientId = process.env.AUTH_SPOTIFY_ID;
	const clientSecret = process.env.AUTH_SPOTIFY_SECRET;

	try {
		// Request token from Spotify API
		const { data } = await axios.post("https://accounts.spotify.com/api/token", null, {
			headers: {
				"Content-Type": "application/x-www-form-urlencoded",
			},
			params: {
				grant_type: "client_credentials",
				client_id: clientId,
				client_secret: clientSecret,
			},
		});

		const { access_token, expires_in }: TokenData = data;

		const response = NextResponse.json({ message: "Ok", status: 200 });
		response.cookies.set("access_token", access_token, { path: "/" });
		return response;
	} catch (error) {
		console.error("Error:", error);
		return NextResponse.json({ error: "Failed to fetch access token" });
	}
}
```
I can't understand why my "access_token" cookie is not being set when I make that axios request in ```localhost:3000/dashboard```. It is only set when I hit ```localhost:3000/api/dashboard```

### Additional information

```js
**Environment**:
* Next v14.2.13
* axios v1.7.7
```

**Comment by icyJoseph** [comment, marked as answer]
Hi,

The problem here is that, `await axios.get("http://localhost:3000/api/dashboard");`, is making its own request, out to the internet, and back to your route. That is to say, it has its own Request and Response objects, independent of the Request and Response, that the user requesting your page have started.

If you assign the response of `await axios.get("http://localhost:3000/api/dashboard");` to a value, and then look for the headers->cookies, you'll most likely see the cookie you set in your GET handler.

Moreover, this is an anti-pattern, you shouldn't:

- reach for API/Route Handlers, from your render pass, see more here, https://nextjs-faq.com/fetch-api-in-rsc
- in App Router, because of how HTTP is designed, the outgoing cookies are set before streaming begins, so when you are rendering your components, setting cookies is not possible
